pytorch_lightning/plugins/pipe_rpc_plugin.py

```diff
@@ -0,0 +1,315 @@
+        os.environ["MASTER_PORT"] = "15000"
+        rpc.init_rpc(f"worker{global_rank}", rank=global_rank, world_size=world_size)
+
+        self.num_gpus_per_model = len(self.balance)
+        ensure_divisibility(world_size, self.num_gpus_per_model)
+        num_model_parallel = self.num_gpus_per_model / world_size
+        mpu.initialize_model_parallel(num_model_parallel, world_size)
+        self._broadcast_and_barrier_supported = num_model_parallel > 1
+
+        automatic_optimization = trainer.train_loop.automatic_optimization
+        if automatic_optimization:
+            raise MisconfigurationException(
+                'PipePlugin is currently not supported in automatic optimization')
+
+        if trainer.amp_backend is not None:
+            raise MisconfigurationException(
+                'PipePlugin is currently not supported in Automatic Mixed Precision')
+
+        self.trainer = trainer
+        model = trainer.get_model()
+
+        if global_rank != 0:
+            # For RPC, all ranks other than 0 just need to call rpc.shutdown()
+            torch_distrib.barrier()
+            rpc_pipe.PipeModel.trainer = model.trainer
+            rpc_pipe.PipeModel.configure_optimizers = model.configure_optimizers
+            torch.distributed.rpc.shutdown()
+            return True
+
+        # Create pipe_module
+        model = trainer.get_model()
+        self._find_pipe_module(model)
+        torch_distrib.barrier()
+        model.foreach_worker(register_optimizers, include_self=True)
+        trainer.is_master = True
+
+    def init_ddp_connection(
+            self,
+            trainer,
+            cluster_environment,
+            global_rank: int,
+            world_size: int,
+            is_slurm_managing_tasks: bool = True,
+    ) -> None:
+
+        skip = self.pre_init_ddp_connection(trainer)
+        if skip:
+            return os.getenv("LOCAL_RANK") != 0
+
+        super().init_ddp_connection(
+            trainer=trainer,
+            cluster_environment=cluster_environment,
+            global_rank=global_rank,
+            world_size=world_size,
+            is_slurm_managing_tasks=is_slurm_managing_tasks
+        )
+
+        return self.init_rpc_connection_and_pipe(trainer, global_rank, world_size)
+
+    def create_optimizers_map(self):
+        self._optimizers_map = {opt_idx: False for opt_idx, opt in enumerate(self.trainer.optimizers)}
+
+    def configure_ddp(
+            self, model: LightningModule, device_ids: List[int]
+    ) -> DistributedDataParallel:
+        self.create_optimizers_map()
+        ddp_plugin = RPCPlugin(process_group=mpu.get_data_parallel_group()).configure_ddp(model, device_ids)
+        return ddp_plugin
+
+    def _save_model(self, checkpoint_save_model, last_filepath, trainer, pl_module):
+        automatic_optimization = self.trainer.train_loop.automatic_optimization
+        model = self.trainer.get_model()
+        if not automatic_optimization:
```

Comment:
Could you explain to me why saving the model can only happen in manual optimization?

Reply:
But we don't support automatic_optimization yet. It is not tested, but no reason outside of that.